pysite/views/wiki/edit.py

```diff
@@ -56,4 +91,76 @@ def post(self, page):
             conflict="replace"
         )
 
+        after = obj['rst'].splitlines(keepends=True) or []
+        if len(after) == 0:
+            return redirect(url_for("wiki.edit", page=page), code=303)
+
+        if not after[-1].endswith("\n"):
+            after[-1] += "\n"  # Does the same thing as L57
+
+        diff = difflib.unified_diff(before, after, fromfile="before.rst", tofile="after.rst")
+        diff = "".join(diff)
+
+        gist_payload = {
+            "description": f"Changes to: {obj['title']}",
+            "public": False,
+            "files": {
+                "changes.md": {
+                    "content": f"```diff\n{diff}\n```"
+                }
+            }
+        }
+
+        headers = {
+            "Authorization": f"token {GITHUB_TOKEN}",
+            "User-Agent": "Discord Python Wiki (https://github.com/discord-python)"
+        }
+
+        gist = requests.post("https://api.github.com/gists",
+                             json=gist_payload,
+                             headers=headers)
+
+        audit_payload = {
+            "username": "Wiki Updates",
+            "embeds": [
+                {
+                    "title": "Page Edit",
+                    "description": f"**{obj['title']}** was edited by **Joseph**"
+                                   f".\n\n[View diff]({gist.json().get('html_url')})",
+                    "color": 4165079,
+                    "timestamp": datetime.datetime.utcnow().isoformat(),
+                    "thumbnail": {
+                        "url": "https://pythondiscord.com/static/logos/logo_discord.png"
+                    }
+                }
+            ]
+        }
+
+        if WIKI_AUDIT_WEBHOOK:
+            requests.post(WIKI_AUDIT_WEBHOOK, json=audit_payload)
+
+        # Add the post to the revisions table
+        revision_payload = {
+            "slug": page,
+            "post": obj,
+            "date": datetime.datetime.utcnow().timestamp(),
+            "user": self.user_data.get("user_id")
+        }
+
+        del revision_payload["post"]["slug"]
+
+        self.db.insert(self.revision_table_name, revision_payload)
+
         return redirect(url_for("wiki.page", page=page), code=303)  # Redirect, ensuring a GET
+
+    def patch(self, page):
+        current = self.db.get(self.table_name, page)
+        if current.get("lock_expiry"):
+            if current["lock_user"] != self.user_data.get("user_id"):
+                return "", 400
+            new_lock = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)
+            self.db.insert(self.table_name, {
+                "slug": page,
+                "lock_expiry": new_lock.timestamp()
```

Comment:
Doesn't rethinkdb support actual datetimes?

Reply:
It has it's own implementation, [documented here](https://rethinkdb.com/api/python/time/), but it's not very clear about how you can convert between the two, there are methods which get it to iso time but iso to python datetime is also tricky. I can look more into it if needed, but the current datetime implementation is nicer to work with in my opinion (we get things like time deltas).